src/exchanges/populate.ts

```diff
@@ -0,0 +1,331 @@
+
+          const types = getTypes(schema, typeInfo);
+          const newSelections = types.reduce((p, t) => {
+            const typeFrags = typeFragments[t.name];
+            if (!typeFrags) {
+              return p;
+            }
+
+            return [
+              ...p,
+              ...typeFragments[t.name].map(({ fragment }) => {
+                // Add used fragment for insertion at Document node
+                getUsedFragments(fragment).forEach(f =>
+                  requiredUserFragments.add(userFragments[f])
+                );
+
+                // Add fragment for insertion at Document node
+                additionalFragments = {
+                  ...additionalFragments,
+                  [fragment.name.value]: fragment,
+                };
+
+                return {
+                  kind: 'FragmentSpread',
+                  name: {
+                    kind: 'Name',
+                    value: fragment.name.value,
+                  },
+                } as const;
+              }),
+            ];
+          }, [] as FragmentSpreadNode[]);
+
+          const existingSelections =
+            (node.selectionSet && node.selectionSet.selections) || [];
+          const selections =
+            existingSelections.length + newSelections.length !== 0
+              ? [...newSelections, ...existingSelections]
+              : [
+                  {
+                    kind: 'Field',
+                    name: {
+                      kind: 'Name',
+                      value: '__typename',
+                    },
+                  },
+                ];
+
+          return {
+            ...node,
+            directives,
+            selectionSet: {
+              kind: 'SelectionSet',
+              selections,
+            },
+          };
+        },
+      },
+      Document: {
+        leave: node => {
+          return {
+            ...node,
+            definitions: [
+              ...node.definitions,
+              ...Object.values(additionalFragments),
+              ...[...requiredUserFragments],
+            ],
+          };
+        },
+      },
+    })
+  );
+};
+
+/** Get all possible types for node with TypeInfo. */
+const getTypes = (schema: GraphQLSchema, typeInfo: TypeInfo) => {
+  const type = typeInfo.getType();
+
+  if (!type || !('ofType' in type)) {
+    console.warn('PopulateExchange: Unsupported type at populate decorator.');
+    return [];
+  }
+
+  const ofType = type.ofType;
+
+  if (ofType instanceof GraphQLInterfaceType) {
+    return schema.getPossibleTypes(ofType);
+  }
+
+  return [ofType as GraphQLObjectType];
+};
+
+/** Get fragment names referenced by node. */
+const getUsedFragments = (node: ASTNode) => {
+  let names: string[] = [];
+
+  visit(node, {
+    FragmentSpread: f => {
+      names = [...names, f.name.value];
+    },
+  });
+
+  return names;
+};
+
+const getType = (t: TypeInfo) => {
+  const type = t.getType() as any;
+  return type.ofType || type.toString();
```

Comment:
This seems interesting; I get that this could be almost anything (https://github.com/graphql/graphql-js/blob/c7f11f3b95367247cc4ce3adb4c24313d2943c5b/src/type/definition.d.ts#L100-L114) but what are we trying to get here with `toString()`?

Reply:
This should really be

```suggestion
  return type.ofType.toString() || type.toString();
```

FYI:
 - ofType = List/NonNull
 - toString = Object/Interface/string

Reply:
Edit

```suggestion
  return (type.ofType && type.ofType.toString()) || type.toString();
```